### PR TITLE
blockstore: atomize slot clearing, relax parent slot meta check 

### PR DESCRIPTION
**Pull request by AshwinSekar**
clear_unconfirmed_slot can leave blockstore in an irrecoverable state
if it panics in the middle. write batch this function, so that any
errors can be recovered after restart.

### CODE REVIEW

File: ledger/src/blockstore.rs
Comment (AshwinSekar) [reply ledger/src/blockstore.rs:1201]: the parent meta is cleaned up after `run_purge` so the child's `slot_meta` no longer exists. later `slot_meta.clear_unconfirmed_slot()` initialized the child's `slot_meta` as an orphan before inserting it.

`clear_parent_next_slot()` will only error if any of the blockstore functions fail, since we have:
`if let Some(mut parent_slot_meta) = parent_slot_meta`, we won't panic if the parent's meta is not found.

Comment (AshwinSekar) [reply ledger/src/blockstore.rs:1202]: essentially i believe we should still panic later if:
- `blockstore.meta(parent_slot).is_err()`
- `blockstore.meta_cf.put(parent_slot, &parent_slot_meta).is_err()`
but not if
- `blockstore.meta(parent_slot).unwrap().is_none()`

File: ledger/src/blockstore.rs
Comment (steviez) [reply ledger/src/blockstore.rs]: > Also clearing the parent's next slots will leave this in a worst state if the shreds aren't also cleaned up, because repair won't happen

We could update `run_purge` to take a flag, and perform this cleanup of next_slots too. The benefit there would be that we could tie into the same `WriteBatch` to make the operation truly atomic (all succeeds or all fails, based on the write batch commit success)

Comment (AshwinSekar) [reply ledger/src/blockstore.rs]: I'm not sure if `ReplayStage` should try to recover from any of the current panics here. It seems that if we have a rocksdb error here we should panic as blockstore is in a bad state.
I think the motivation is that if the rocksdb error happens to be during the `parent_slot_meta` cleanup, we should still reinsert the `child`'s `slot_meta` as upon restart we might be able to squeak by.

in retrospect it seems brittle to rely on this behavior so perhaps the change should be to only relax the parent's `slot_meta` cleanup and still fail fast on any rocksdb error.

Comment (AshwinSekar) [reply ledger/src/blockstore.rs]: I like the idea of extending the use of `WriteBatch` to cover this whole function, it seems brittle to rely on specific orderings and recoveries, and better to say this function should be atomic.

**Comment by AshwinSekar** [comment]
reworked as per https://github.com/solana-labs/solana/pull/35124#discussion_r1481923677
the writes now happen atomically

File: ledger/src/blockstore/blockstore_purge.rs
Comment (steviez) [inline ledger/src/blockstore/blockstore_purge.rs:290]: We don't have to do this sanity checking if we fetch the `SlotMeta` in this function ourselves instead of relying on the caller 

Technically, this would add some overhead since we'd be fetching the `SlotMeta` twice, once here and once in `Blockstore::clear_unconfirmed_slot()`. I think that's ok, but if we're conscious about that, I think we could remove the `SlotMeta` check in `Blockstore::clear_unconfirmed_slot()` anyways. The rocksdb range deletes are cheap, and this function should only be getting called if `ReplayStage` knows it has something that it needs to delete

File: ledger/src/blockstore/blockstore_purge.rs
Comment (steviez) [inline ledger/src/blockstore/blockstore_purge.rs:194]: Instead of another function / more arguments, how about adding a new case to the `PurgeType`. Maybe something like `ExactAndCleanupChaining`. We already have a spot where we switch on `purge_type` in `Blockstore::run_purge_with_stats()` so I think this would tie in nicely.

Comment (steviez) [reply ledger/src/blockstore/blockstore_purge.rs:194]: Actually, I don't think we need a separate type; I think we should just update the logic in the `Exact` case. The non-test case callers that use `PurgeType::Exact` are:
- https://github.com/solana-labs/solana/blob/eeb0cf1ea8db812e1ac65571962f6b0e918dabba/core/src/validator.rs#L2189-L2190
- https://github.com/solana-labs/solana/blob/eeb0cf1ea8db812e1ac65571962f6b0e918dabba/ledger/src/blockstore.rs#L1177-L1178
- https://github.com/solana-labs/solana/blob/eeb0cf1ea8db812e1ac65571962f6b0e918dabba/ledger/src/blockstore/blockstore_purge.rs#L84-L85
    - This code path only reached from https://github.com/solana-labs/solana/blob/eeb0cf1ea8db812e1ac65571962f6b0e918dabba/ledger-tool/src/blockstore.rs#L942-L948
    
The case we wanted to update is the second bullet, and we absolutely want to fixup chaining for that case. The other two cases are both accompanied by calls to `Blockstore::purge_from_next_slots()` which is essentially doing the same logic as `Blockstore::clear_unconfirmed_slot()` in cleaning up the chaining.

So, It seems we can condense more code

Comment (carllin) [reply ledger/src/blockstore/blockstore_purge.rs:194]: `purge_from_next_slots()` looks like it would have pretty terrible performance: https://github.com/solana-labs/solana/blob/ba189b14542f96d63a9b2f7a2588f49d65c74f12/ledger/src/blockstore/blockstore_purge.rs#L98-L102. It starts iterating from slot 0 👻

Comment (steviez) [reply ledger/src/blockstore/blockstore_purge.rs:194]: Yeah, the function definitely needs some work. Consider the following graph:
```
4 - 5 - 6 - 9 - x
 \
  7 - 8 - x
```  
If we call with `from_slot = 6, to_slot = 9`, we'd expect the following afterwards:
```
4 - 5 - x
```
But, if we start with `from_slot`'s (`6`) parent (`5`), `4` would still have `7` in its' `next_slots` despite `7` having been purged.

So, I agree that we can iterate `[from_slot, to_slot]` instead of `[0, to_slot]`. But. for every slot in that range, I think we need to update the parent's slot, similar to what `clear_unconfirmed_slot()` is currently doing. So, overhead will look pretty similar for the `clear_unconfirmed_slot()` case where `from_slot == to_slot`.

Comment (steviez) [reply ledger/src/blockstore/blockstore_purge.rs:194]: Another thing that just occurred to me - by moving the functionality within the same write batch, we should ensure that the ordering in write batch is respected.

Suppose we have this chain:
```
4 - 5 - 6
```
If we call with `from_slot = 5, to_slot = 5`, there should still be an empty `SlotMeta` for `5` that has `6` in its' `next_slots` vector.

But, earlier in the function, we will have range deleted `[5, 5]`. So, we need to ensure that the `SlotMeta` update comes after the range delete AND double check that ordering is ensured (ie the `SlotMeta` for `5` still present when the function finishes. I believe this to be true, but again, just calling it out since I thought of it.

The logic here seems tricky enough that a unit test might be in order too ... sorry Ashwin, no good deed goes unpunished. Maybe we can whiteboard this out in the next week to make sure we cover all the edges

**Comment by AshwinSekar** [comment]
In the interest of keeping the backport surface small, i've added a new `PurgeType` variant, and fetched the `slot_meta` within the purge function. In a future master PR I can unify the remaining `PurgeType::Exact` callers as suggested here https://github.com/solana-labs/solana/pull/35124#discussion_r1483578012 and remove this variant.

File: ledger/src/blockstore/blockstore_purge.rs
Comment (carllin) [inline ledger/src/blockstore/blockstore_purge.rs:288]: This and the above together seem like enough unique logic to warrant a separate function. I think it would make sense to:

1.  move this into a separate function that takes a write_batch and a range https://github.com/solana-labs/solana/blob/45e292d8214f84972c7cc798a43190ec6cbf1c0c/ledger/src/blockstore/blockstore_purge.rs#L186-L253
2. Make a `purge_single_slot(slot: slot)` function that makes a write batch and moves it into 1.

Comment (AshwinSekar) [reply ledger/src/blockstore/blockstore_purge.rs:288]: I was planning to bring this up for a future pr: I think actually this logic should be always be present? 
Right now what we're doing is ensuring that if we purge a single slot from a fork, the repair of that slot will allow replay to smoothly process the fork. 

However for a regular purge:
`4 - 5 - 6 - 7 - 8 - 9 - 10`
if we purge `Exact` for `from_slot = 5, to_slot = 8` then I don't believe we'll ever be able to replay slots `9, 10` on a successful repair. 
it seems we should extend the logic so that:
```
 parent_slot_meta
      .next_slots
      .retain(|&next_slot| next_slot != slot);
```
is **always** run on `from_slot` and:
```
slot_meta.clear_unconfirmed_slot();
write_batch.put::<cf::SlotMeta>(slot, &slot_meta)?;
```
is **always** run on `to_slot`.

Comment (AshwinSekar) [reply ledger/src/blockstore/blockstore_purge.rs:288]: but perhaps this is out of scope for this function and we should only `run_purge` for rooted slots. in that case I agree with carl's idea of separating this out to a different function that is intended for forks currently being replayed.

File: ledger/src/blockstore/blockstore_purge.rs
Comment (carllin) [inline ledger/src/blockstore/blockstore_purge.rs:267]: does it make sense to include `purge_special_columns_exact()` in here?

```
fn purge_range() {}
fn purge_range_special_columns(){}
fn do_purge_range(&self, write_batch: &mut WriteBatch, from_slot: Slot, to_slot: Slot, should_purge_special_columns: bool)
```

File: ledger/src/blockstore/blockstore_purge.rs
Comment (carllin) [inline ledger/src/blockstore/blockstore_purge.rs:179]: hmmm it could also be because the parent is just really old because somebody generated this maliciously right? Think it's fine to leave out the cluster restart scenario.

File: ledger/src/blockstore/blockstore_purge.rs
Comment (carllin) [inline ledger/src/blockstore/blockstore_purge.rs:167]: let's make a note here to hard unwrap once lowest cleanup slot is dependable, would be much simpler

File: ledger/src/blockstore/blockstore_purge.rs
Comment (AshwinSekar) [reply ledger/src/blockstore/blockstore_purge.rs]: good idea, i just passed through `PurgeType` instead of separate functions.

File: ledger/src/blockstore/blockstore_purge.rs
Comment (AshwinSekar) [reply ledger/src/blockstore/blockstore_purge.rs:166]: added it to the else clause

File: ledger/src/blockstore/blockstore_purge.rs
Comment (steviez) [inline ledger/src/blockstore/blockstore_purge.rs:193]: nit: I've hit this case before where I wanted to print a log message locally but still raise the error. After looking at docs real quick, it looks like the following might simplify this (new feature in Rust 1.76):
```rust
self.db.write(write_batch)
    .inspect_err(|e| error!("Error {:?} ....))?;
```
Also, I don't think that `retrying` bit in the message is valid here (or in the other spot) so can you please remove it

File: ledger/src/blockstore/blockstore_purge.rs
Comment (steviez) [inline ledger/src/blockstore/blockstore_purge.rs:229]: Similar to other comment:
- I think you might be able to simplify slightly with `.inspect_err()` + `?`
- That `retrying` bit isn't accurate, so can you adjust the message. Can you also adjust the message to include `from_slot` and `to_slot`

**Review by steviez — COMMENTED** [review]
Mostly nits, liking the way this is coming together. Will take a final pass shortly

File: ledger/src/blockstore/blockstore_purge.rs
Comment (AshwinSekar) [reply ledger/src/blockstore/blockstore_purge.rs]: nice, `inspect_err` is slick.

File: ledger/src/blockstore.rs
Comment (steviez) [inline ledger/src/blockstore.rs:1168]:
```suggestion
        // Shreds inherently know their parent slot, and a parent's SlotMeta `next_slots` list
        // will be updated when the child is inserted (see `Blockstore::handle_chaining()`).
```
I think this reads a little better, what do you think ?

File: ledger/src/blockstore.rs
Comment (steviez) [inline ledger/src/blockstore.rs:1169]:
```suggestion
        // However, we are only purging and repairing the parent slot here. Since the child will not be
```
nit

File: ledger/src/blockstore/blockstore_purge.rs
Comment (steviez) [inline ledger/src/blockstore/blockstore_purge.rs:150]:
```suggestion
    /// Additionally, we cleanup the parent of `slot` by clearing `slot` from
```
nit

File: ledger/src/blockstore/blockstore_purge.rs
Comment (steviez) [inline ledger/src/blockstore/blockstore_purge.rs:161]: nit: Think we can use `?` instead of the explicit `.expect()` and let `clear_unconfirmed_slot()` decide what to do with `Err(...)` (still a panic but if we decide to bubble errors of later it will be less refactor then)

File: ledger/src/blockstore/blockstore_purge.rs
Comment (steviez) [inline ledger/src/blockstore/blockstore_purge.rs:216]: nit: I know it was like this before, but can you update the above `.expect()` to use `?` as well

File: ledger/src/blockstore/blockstore_purge.rs
Comment (steviez) [inline ledger/src/blockstore/blockstore_purge.rs:183]:
```suggestion
        // Retain a SlotMeta for `slot` with the `next_slots` field retained
```
We already provide some justification for why we're doing this in the `///` comments above the function

**Review by steviez — COMMENTED** [review]
Just a few more nits on comments + simplification. Think we're good for a ship-it after these changes; I can give that quickly after you make the changes but think it'd be nice to get a final pass & approval from Carl too

**Review by steviez — APPROVED** [review]
LGTM, thanks for pushing through after we collectively changed our mind several times. Just to be sure, let's see if we can get @carllin to give one more pass too

File: ledger/src/blockstore/blockstore_purge.rs
Comment (carllin) [inline ledger/src/blockstore/blockstore_purge.rs:253]: nit: I think it's cleaner  encapsulation to pass a `should_purge_special_columns` boolean here, and then have a function `PurgeType::should_should_purge_special_columns() -> bool` method on `PurgeType`. This way `purge_slot_cleanup_chaining()` doesn't have to know about `PurgeType` at all

Comment (steviez) [reply ledger/src/blockstore/blockstore_purge.rs:253]: > This way purge_slot_cleanup_chaining() doesn't have to know about PurgeType at all

I see the merit of adding the new function `purge_slot_cleanup_chaining()` as this has the unique functionality from `run_purge_with_stats`. However, I'm not sure I see the benefit of the boolean over the enum; can you elaborate?

The enum is part of the public API, so I think it is reasonable to expect someone to know about it. And the `Exact` value of the enum means "go purge the special columns for these slots right now", which is what the boolean would convey.

Comment (carllin) [reply ledger/src/blockstore/blockstore_purge.rs:253]: 1. It seems like the `PurgeType` was designed specifically for the `run_purge_with_stats` workflow, so it seemed better to keep it isolated there rather than mixing it with this new clean function
2. `purge_slot_cleanup_chaining` as a utility function seems like it should just decide which columns to clean. Seems like cramming the `PurgeType` in there was too high level.

**Pull request merged**
